design/permissions.md

```diff
@@ -0,0 +1,181 @@
+			"items": {
+				"type": "string",
+				"enum": ["user-tracking", "user-interaction"]
+			}
+		},
+		"optionalPermissions": {
+			"type": "array",
+			"items": {
+				"type": "string",
+				"enum": ["user-tracking", "user-interaction"]
+			}
+		}
+	}
+}
+```
+
+Permissioning flow
+--------------------
+
+Legend:
+
+* [App] Step executed by MRE developer code
+* [SDK] Step executed by MRE SDK
+* [Runtime] Step executed in client-side MRE runtime
+* [Host] Step executed by host application
+
+Startup:
+
+1. [App] During development, the app developer authors a static manifest file, or uses the WebHost API to generate one.
+2. [Runtime|Host] Initializes the MRE API with a permissions manager instance that will receive new permission requests.
+	If permission decisions are persistent, load them into memory now. Default provided implementation does not persist.
+3. [Host] The host wishes to run an MRE, so creates an IMixedRealityExtensionApp instance and calls Startup().
+4. [Runtime] Downloads the manifest from the provided MRE server. If missing, assumes no permissions required.
+5. [Runtime] Calls into the permission manager requesting the manifest-listed permissions.
+6. [Runtime|Host] Permission manager evaluates the required and optional MRE permissions against the current set of
+	grants/denials for the MRE's origin, and determines if any new permission decisions need to be made by the user.
+	If so:
+	1. [Host] Present the choices to the user by whatever means the host sees fit, persist the decision if desired,
+		and report the result to the Runtime. Optional permissions should be presented in such a way as they can be
+		granted or denied individually, but required permissions should be decided as a group.
+7. [Runtime] Determine if the MRE has sufficient permission to run, i.e. `execution` and all the required permissions
+	are granted. If not, abort connecting to the MRE.
+8. [Runtime] Startup proceeds like normal, but if a `user-joined` message is sent, the user payload must include
+	the IDs of any permissions that were granted to this MRE by this user (other than `execution`, which is implied).
+
+
+Permissions Manager
+---------------------
+
+Host apps must provide hooks for the MRE subsystem to obtain permission from users. This is provided to `MREAPI.InitializeAPI`
+as an implementation of `IPermissionManager`, which has the following methods:
+
+* `Task<Permissions> PromptForPermission(...)` - Request permissions from the user, and return a Task that resolves
+	with those permissions the user has granted. Takes the following arguments:
+	* `string appOrigin` - The origin of the MRE requesting permission, as described above.
+	* `string appDisplayName` - A human-readable identifier for the MRE server provided from the manifest.
+	* `Permissions permissionsNeeded` - A bitfield of the permissions required to run the app.
+	* `Permissions permissionsWanted` - A bitfield of permissions that the app can use, but are not needed to run.
```

Comment:
Since this is something being passed to a host app, we may want to consider passing this in a more friendly unpacked format, so that the host app can easily digest this for their UI.  Otherwise the parsing of the bitfield in to actionable items and repacking has to happen as code written by each host app.  If we send this as something already parsed and receive in the same unpacked manner, we can abstract away the need to deal with separating out actionable items from a bitfield. Maybe an array of Permissions enum value that can easily be iterated over to add prompts to an overall UI?

Reply:
In fact this is how we transport them in the json as an array of strings, so you could just wait until the permissions task has completed before you pack this in to a bitfield of approved permissions.

---
In reply to: [441643214](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441643214) [](ancestors = 441643214)

Reply:
That's a good point: since we're sending the perms down as arrays, there's no point in repacking them. I like bitfields because of O(1) lookups, but for <32 items I suppose it doesn't matter.

---
In reply to: [441644095](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441644095) [](ancestors = 441644095,441643214)

Reply:
I mean still packing for lookups is fine, I was just suggesting doing that packing after the host app has gathered the approved permissions.

Reply:
I'm pretty sure C# lets you do foreach loops on bitfields of enums. It's cool like that. Might not matter, even for enumeration. I'll play with this come implementation time.

---
In reply to: [441733983](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441733983) [](ancestors = 441733983)

Reply:
It's a bit convoluted how you have to iterate over the values of the enum, but there is a way.  It is still easier at this point since the message is carrying as an array of values to just pass that array in to the dialog and receive one back.

---
In reply to: [441766267](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441766267) [](ancestors = 441766267,441733983)